src/H5FDros3.c

```diff
@@ -599,6 +608,240 @@ H5FD__ros3_fapl_free(void *_fa)
     FUNC_LEAVE_NOAPI(SUCCEED)
 } /* end H5FD__ros3_fapl_free() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5Pget_fapl_ros3_token
+ *
+ * Purpose:     Returns session/secutiry token of the ros3 file access
+ *              property list though the function arguments.
+ *
+ * Return:      Success:        Non-negative
+ *
+ *              Failure:        Negative
+ *
+ * Programmer:  Jan-Willem Blokland
+ *              2023-05-26
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Pget_fapl_ros3_token(hid_t fapl_id, size_t size, char *token_dst /*out*/)
+{
+    H5P_genplist_t *plist                                   = NULL;
+    char            token_src[H5FD_ROS3_MAX_SECRET_TOK_LEN] = {0};
+    htri_t          token_exists;
+    size_t          tokenlen  = 0;
+    herr_t          ret_value = SUCCEED;
+
+    FUNC_ENTER_API(FAIL)
+    H5TRACE2("e", "i*s", fapl_id, token);
+
+#if ROS3_DEBUG
+    HDfprintf(stdout, "H5Pget_fapl_ros3_token() called.\n");
+#endif
+
+    if (size == 0)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "size cannot be zero.")
+    if (token_dst == NULL)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "token_dst is NULL")
+
+    if (NULL == (plist = H5P_object_verify(fapl_id, H5P_FILE_ACCESS)))
+        HGOTO_ERROR(H5E_PLIST, H5E_BADTYPE, FAIL, "not a file access property list")
+    if (H5FD_ROS3 != H5P_peek_driver(plist))
+        HGOTO_ERROR(H5E_PLIST, H5E_BADVALUE, FAIL, "incorrect VFL driver")
+    if ((token_exists = H5P_exist_plist(plist, ROS3_TOKEN_PROP_NAME)) < 0)
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "failed to check if property token exists in plist")
+    if (token_exists) {
+        if (H5P_get(plist, ROS3_TOKEN_PROP_NAME, token_src) < 0)
```

Comment:
Should be `&token_src` in this case since all the callbacks are expecting to receive a `void **` that gets converted to a `char **`

Reply:
This comment and comment of the function `H5FD__ros3_str_token_copy()`, `H5FD__ros3_str_token_close()` and `H5FD__ros3_str_token_delete()` are more or less related to each other.

As you noticed, my current implementation is slightly different compared to code snippets you posted on the forum. The main reason for it is that I want the HDF5 library to store specified token string instead of the pointer to pointer. I assume that is done when calling `H5Pset_ros3_fapl()` for the strings inside the struct `H5FD_ros3_fapl_t`. If this is not true, I need to rethink certain things when implementing ROS3 functionality on my library by making use of HDF5. My aim is that my implementation behaves similar as the existing implementation regarding specifying the security credentials.

Furthermore, specifying the token is optional and if the token is not specified no additional memory will be allocated.

Reply:
Having said that, if you plan to use the implementation where the whole value is stored, you can probably just remove the `H5FD__ros3_str_token_close` and `H5FD__ros3_str_token_delete` callbacks and pass `NULL` for them in the call to `H5Pinsert2`.